### PR TITLE
Continuous batching refactor

### PR DESCRIPTION
**Pull request by remi-or**
This PR refactors the files related to continuous batching and changes a few parts of the logic:
- the way the `max_batch_tokens` and `num_blocks` is computed;
- the static inputs of the CB are now sliced to avoid passing an arbitrary amount of padding in the forward
- the example file for CB has been reworked to allow for more parametrization

### CODE REVIEW

File: examples/pytorch/continuous_batching.py [hunk 1]
```diff
@@ -7,108 +25,232 @@
...
-        model_id,
-        attn_implementation="paged_attention|kernels-community/flash-attn",
-        torch_dtype=torch.bfloat16,
+
+def generate_simple(
+    attn_implementation: str, simple_batch_inputs: list[int], generation_config: GenerationConfig
+) -> list[str]:
+    attn_implementation = {
+        "sdpa_paged": "sdpa",
+        "eager_paged": "eager",
```
Comment (ArthurZucker) [inline examples/pytorch/continuous_batching.py:36]: let's add flash as well

File: examples/pytorch/continuous_batching.py
Comment (ArthurZucker) [inline examples/pytorch/continuous_batching.py:1]: as this increases the complexity, let's add another file for the simple simple usecase

File: src/transformers/generation/continuous_batching/cache.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+import torch
+
+from ...configuration_utils import PretrainedConfig
+from ...generation.configuration_utils import GenerationConfig
+from ...utils.metrics import attach_tracer, traced
+from .core import RequestState, get_device_and_memory_breakdown, logger
+
+
+T = TypeVar("T")
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cache.py]: no 100% sure we need this

File: src/transformers/generation/continuous_batching/cache.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+        # Reshape cache for easier indexing
+        total_slots = self.num_blocks * self.block_size
+        k_cache_flat = self.key_cache[layer_idx].view(self.num_key_value_heads, total_slots, self.head_dim)
+        v_cache_flat = self.value_cache[layer_idx].view(self.num_key_value_heads, total_slots, self.head_dim)
+        k_cache_flat[:, write_index, :] = key_states[0]
+        v_cache_flat[:, write_index, :] = value_states[0]
+        return k_cache_flat[None, :, read_index, :], v_cache_flat[None, :, read_index, :]
+
+
+class PagedAttentionMemoryHandler:
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cache.py:201]: you had a cool computation let's add it as a comment here as well!

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 4]
```diff
@@ -763,49 +110,64 @@ def setup_static_tensors(self):
...
-            "position_ids": self.position_ids,
-            "attention_mask": self.attention_mask,
-            "cu_seq_lens_q": self.cumulative_seqlens_q,
-            "cu_seq_lens_k": self.cumulative_seqlens_k,
-            "write_index": self.write_index,
-            "read_index": self.read_index,
-            "logits_indices": self.logits_indices,
+            "input_ids": self.input_ids[:, :t],
+            "position_ids": self.position_ids[:, :t],
+            "attention_mask": self.attention_mask[:, :, :t, :c], # NOTE: this is probably not used for paged attention
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/continuous_api.py]: In that case fix me if paged then don't allocate

File: src/transformers/generation/continuous_batching/core.py
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/core.py:1]: naming can be improved for the file like "classes.py` or something

**Review by ArthurZucker — APPROVED** [review]
nice 🚀

File: src/transformers/generation/continuous_batching/cache.py (hunk 2 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cache.py]: We dont, removed it

File: src/transformers/generation/continuous_batching/cache.py (hunk 3 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cache.py:201]: done (and reworked it a bit as a factor was missing)

File: examples/pytorch/continuous_batching.py
Comment (remi-or) [reply examples/pytorch/continuous_batching.py:1]: done

File: examples/pytorch/continuous_batching.py (hunk 1 above)
Comment (remi-or) [reply examples/pytorch/continuous_batching.py:36]: Done with `"flash_paged": "flash_attention_2",`

File: src/transformers/generation/continuous_batching/core.py
Comment (remi-or) [reply src/transformers/generation/continuous_batching/core.py:1]: classes.py it is

File: src/transformers/generation/continuous_batching/continuous_api.py (hunk 4 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/continuous_api.py]: Good catch, done

**Review by ArthurZucker — APPROVED** [review]
Let's go 🚀
